Public/Src/Tools/Tool.MsBuildGraphBuilder/MsBuildPredictionCollector.cs

```diff
@@ -0,0 +1,119 @@
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.ContractsLight;
+using System.IO;
+using Microsoft.Build.Prediction;
+
+namespace MsBuildGraphBuilderTool
+{
+    /// <summary>
+    /// Collects predictions from the MSBuild prediction library.
+    /// </summary>
+    /// <remarks>
+    /// This validates the predicted inputs and outputs, in case the predictor is not working properly
+    /// </remarks>
+    public sealed class MsBuildPredictionCollector : IProjectPredictionCollector
+    {
+        private readonly ConcurrentQueue<(string predictorName, string failure)> m_predictionFailures;
+
+        private readonly ICollection<string> m_inputFilePredictions;
+
+        private readonly ICollection<string> m_outputFolderPredictions;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MsBuildPredictionCollector"/> class.
+        /// </summary>
+        /// <remarks>
+        /// The provided collections will be added to by this collector as predictions (or prediction errors) happen.
+        /// </remarks>
+        /// <param name="inputFilePredictions">A collection of input file predictions the collector should add to as needed.</param>
+        /// <param name="outputFolderPredictions">A collection of output folder predictions the collector should add to as needed.</param>
+        /// <param name="predictionFailures">A collection of prediction failures the collector should add to as needed.</param>
+        public MsBuildPredictionCollector(
+            ICollection<string> inputFilePredictions,
+            ICollection<string> outputFolderPredictions,
+            ConcurrentQueue<(string predictorName, string failure)> predictionFailures)
+        {
+            Contract.Assert(inputFilePredictions != null);
+            Contract.Assert(outputFolderPredictions != null);
+            Contract.Assert(predictionFailures != null);
+
+            m_inputFilePredictions = inputFilePredictions;
+            m_outputFolderPredictions = outputFolderPredictions;
+            m_predictionFailures = predictionFailures;
+        }
+
+        /// <inheritdoc/>
+        public void AddInputFile(string path, string projectDirectory, string predictorName)
+        {
+            if (!TryValidatePrediction(path, projectDirectory, predictorName, out string absolutePath))
+            {
+                return;
+            }
+
+            m_inputFilePredictions.Add(absolutePath);
+        }
+
+        /// <inheritdoc/>
+        public void AddInputDirectory(string path, string projectDirectory, string predictorName)
+        {
+            if (!TryValidatePrediction(path, projectDirectory, predictorName, out string absolutePath))
+            {
+                return;
+            }
+
+            if (Directory.Exists(absolutePath))
+            {
+                foreach (string file in Directory.EnumerateFiles(absolutePath))

```

Comment:
>Directory [](start = 40, length = 9)

We have improved enumeration helpers in FileUtilities...

Reply:
Good point Danny. When I wrote this we didn't want to depend on bxl utilities, but that changed already...
@david.federman@outlook.com , Danny is talking about FileUtilities.EnumerateFiles

---
In reply to: [294452228](https://github.com/microsoft/BuildXL/pull/480#discussion_r294452228) [](ancestors = 294452228)